### PR TITLE
Disable exercises

### PR DESCRIPTION
**Pull request by FeisEater**

### CODE REVIEW

**Comment by jruo** [comment]
Resolved merge conflicts

File: app/models/exercise.rb [hunk 1]
```diff
@@ -27,6 +27,8 @@ class Exercise < ActiveRecord::Base
     where(course_id: course.id, gdocs_sheet: gdocs_sheet)
   }
 
+  enum disabled_status: [:enabled, :disabled]
```
Comment (mpartel) [inline app/models/exercise.rb]: I'd just make this a boolean, to be consistent with the other code if nothing else.

File: app/controllers/courses_controller.rb [hunk 2]
```diff
@@ -167,6 +167,11 @@ def save_unlocks
     redirect_to manage_unlocks_organization_course_path(@organization, @course), alert: e.to_s
   end
 
+  def manage_exercises
+    authorize! :teach, @organization
```
Comment (jamox) [inline app/controllers/courses_controller.rb]: could be `authorize! :manage_exercises, @organization`

File: app/controllers/exercises_controller.rb [hunk 3]
```diff
@@ -53,4 +53,23 @@ def show
...
+
+    action = params[:commit] == 'Disable selected' ? :disabled! : :enabled!
+    exercise_params = params[:exercise]
+    exercise_names = []
+    exercise_names = exercise_params.map { |k, v| v == '1' ? k : nil }.compact unless exercise_params.nil?
+
+    exercise_names.each do |name|
+      exercise = Exercise.find_by(name: name)
+      exercise.send(action) unless exercise.nil?
+    end
```
Comment (jamox) [inline app/controllers/exercises_controller.rb]: There must be a more effective way for this. Here we are creating N queries to DB.

File: app/views/courses/manage_exercises.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    </tr>
+    </thead>
+    <tbody>
+    <% @exercises.each do |exercise| %>
+      <tr>
+        <td>
+          <%= render partial: 'exercises/list_single', locals: { exercise: exercise } %>
+        </td>
+        <td>
+          <%= check_box_tag "exercise[#{exercise.name}]" %>
```
Comment (jamox) [inline app/views/courses/manage_exercises.html.erb]: isn't there a rails helper for doing this?

File: app/models/exercise.rb (hunk 1 above)
Comment (jruo) [inline app/models/exercise.rb]: Jamo wanted an enum field for disabling courses (https://github.com/testmycode/tmc-server/pull/265#discussion_r31899442) so I thought why not do the same thing with disabling exericses?

File: app/controllers/courses_controller.rb (hunk 2 above)
Comment (jruo) [inline app/controllers/courses_controller.rb]: We are planning to adjust these kind of permissions more in an upcoming PR.

**Comment by jruo** [comment]
Updated the form to use rails helpers & reduced the amount of DB queries needed to disable exercises to half. I don't know if there is a way to update all of them in one query.

File: app/models/exercise.rb (hunk 1 above)
Comment (mpartel) [inline app/models/exercise.rb]: Oh, I guess we disagree with Jamo then :)
Feel free to leave as is.

File: app/controllers/courses_controller.rb [hunk 5]
```diff
@@ -5,7 +5,7 @@
 
 class CoursesController < ApplicationController
   before_action :set_organization
-  before_action :set_course, only: [:show, :edit, :update, :refresh, :manage_deadlines, :save_deadlines, :enable, :disable, :manage_unlocks, :save_unlocks ]
+  before_action :set_course, only: [:show, :edit, :update, :refresh, :manage_deadlines, :save_deadlines, :enable, :disable, :manage_unlocks, :save_unlocks, :manage_exercises]
```
Comment (jamox) [inline app/controllers/courses_controller.rb]: maybe add next to the `:manage_deadlines`

File: app/controllers/exercises_controller.rb [hunk 6]
```diff
@@ -53,4 +53,22 @@ def show
...
+    authorize! :teach, @organization
+
+    action = params[:commit] == 'Disable selected' ? :disabled! : :enabled!
+    exercise_params = params[:course][:exercises]
+    exercise_params.reject!(&:blank?)
+
+    exercises = Exercise.where(id: exercise_params)
+    exercises.each do |ex|
+      ex.send(action)
+    end
```
Comment (jamox) [inline app/controllers/exercises_controller.rb]: not too happy with this each loop. can't we just batch update `Exercise.where(id: exercise_params).update_all(disabled_status: 1)` (or by using the enums)

File: app/controllers/exercises_controller.rb [hunk 7]
```diff
@@ -53,4 +53,22 @@ def show
...
+
+    redirect_to manage_exercises_organization_course_path(@organization, @course),
+                notice: 'Selected exercises successfully updated.'
```
Comment (jamox) [inline app/controllers/exercises_controller.rb]: too much indent

File: app/models/exercise.rb (hunk 1 above)
Comment (jamox) [inline app/models/exercise.rb]: :)

File: app/views/courses/manage_exercises.html.erb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      <th>
+        Select
+      </th>
+    </tr>
+    </thead>
+    <tbody>
+    <%= f.collection_check_boxes(:exercises, @exercises, :id, :name) do |c| %>
+      <tr>
+        <td>
+          <%= render partial: 'exercises/list_single', locals: { exercise: @exercises.detect { |e| e.id == c.value } } %>
```
Comment (jamox) [inline app/views/courses/manage_exercises.html.erb]: this select seems maybe too inefficient

File: app/views/exercises/_list_single.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<%= link_to exercise.name, exercise %>
+<%
+   notes = []
+   notes << 'hidden' if exercise.hidden?
+   notes << 'expired' if exercise.expired_for?(current_user)
+   notes << 'locked' unless exercise.unlocked_for?(current_user)
+   notes << 'disabled' if exercise.disabled?
+%>
+<% unless notes.empty? %>
+  (<%= notes.join(', ') %>)
```
Comment (jamox) [inline app/views/exercises/_list_single.html.erb]: Still wondering if there would be a nicer way for this -- keep in mind users might try to inject additional fields for the array

File: app/controllers/exercises_controller.rb (hunk 7 above)
Comment (jruo) [inline app/controllers/exercises_controller.rb]: Rubocop doesn't allow any other way to indent here. "Align the parameters of a method call if they span more than one line."

**Comment by jruo** [comment]
Updated

**Comment by jamox** [comment]
LGTM, nice work. Due to the github repo to local repo changes a few features were failing; please fix in a new PR.

**Pull request merged**